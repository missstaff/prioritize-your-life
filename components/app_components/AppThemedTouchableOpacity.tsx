import React from "react";
import { ScaledSheet, s } from "react-native-size-matters";
import {
  StyleProp,
  TextStyle,
  TouchableOpacity,
  TouchableOpacityProps,
  ViewStyle,
} from "react-native";
import { AppThemedText } from "./AppThemedText";
import { COLORS } from "@/constants/Colors";

export type ThemedAppThemedTouchableOpacity = TouchableOpacityProps & {
  children: React.ReactNode;
  textStyles?: StyleProp<TextStyle>;
  btnStyles?: StyleProp<ViewStyle>;
  onPress?: () => void;
};

/**
 * Custom TouchableOpacity component for the app.
 * @param {TouchableOpacityProps} props - The TouchableOpacity props.
 * @param {React.ReactNode} props.children - The children to be rendered inside the TouchableOpacity.
 * @param {StyleProp<ViewStyle>} props.textStyles - The styles to be applied to the text inside the TouchableOpacity.
 * @param {StyleProp<ViewStyle>} props.btnStyles - The styles to be applied to the TouchableOpacity.
 * @param {Function} props.onPress - The function to be called when the TouchableOpacity is pressed.
 * @param {boolean} props.disabled - A boolean value to disable the TouchableOpacity.
 * @returns {JSX.Element} - The rendered TouchableOpacity component.
 */
const AppThemedTouchableOpacity = ({
  children,
  onPress,
  textStyles,
  btnStyles,
  disabled
}: ThemedAppThemedTouchableOpacity): JSX.Element => {
  return (
    <TouchableOpacity disabled={disabled} onPress={onPress} style={[styles.button, btnStyles]}>
      <AppThemedText style={[styles.buttonText, textStyles]}>
        {children}
      </AppThemedText>
    </TouchableOpacity>
  );
};

export default AppThemedTouchableOpacity;

const styles = ScaledSheet.create({
  button: {
    alignItems: "center",
    backgroundColor: COLORS.primary,
    borderRadius: s(5),
    marginBottom: s(10),
    padding: s(10),
    width: "80%",
  },
  buttonText: {
    color: COLORS.white,
  },
});
