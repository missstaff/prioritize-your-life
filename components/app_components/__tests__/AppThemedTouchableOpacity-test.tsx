import React from "react";
import renderer from "react-test-renderer";
import AppThemedTouchableOpacity from "../AppThemedTouchableOpacity";
import { TouchableOpacity } from "react-native";

describe("AppThemedTouchableOpacity Tests", () => {
  
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly with default props", () => {
    const tree = renderer
      .create(
        <AppThemedTouchableOpacity onPress={jest.fn()}>
          Click Me
        </AppThemedTouchableOpacity>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders correctly with custom styles", () => {
    const customBtnStyle = { backgroundColor: "red" };
    const customTextStyle = { fontSize: 20 };

    const tree = renderer
      .create(
        <AppThemedTouchableOpacity
          btnStyles={customBtnStyle}
          textStyles={customTextStyle}
        >
          Custom Button
        </AppThemedTouchableOpacity>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("triggers onPress correctly", () => {
    const onPressMock = jest.fn();

    const component = renderer.create(
      <AppThemedTouchableOpacity onPress={onPressMock}>
        Pressable
      </AppThemedTouchableOpacity>
    );

    const button = component.root.findByType(TouchableOpacity);
    button.props.onPress();

    expect(onPressMock).toHaveBeenCalled();
  });

  it("renders correctly when disabled", () => {
    const tree = renderer
      .create(
        <AppThemedTouchableOpacity disabled>
          Disabled Button
        </AppThemedTouchableOpacity>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
