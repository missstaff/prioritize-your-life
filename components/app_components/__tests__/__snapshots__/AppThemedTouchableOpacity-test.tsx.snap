// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppThemedTouchableOpacity Tests renders correctly when disabled 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#CD03EF",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Disabled Button
  </Text>
</View>
`;

exports[`AppThemedTouchableOpacity Tests renders correctly with custom styles 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "red",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          {
            "fontSize": 20,
          },
        ],
      ]
    }
  >
    Custom Button
  </Text>
</View>
`;

exports[`AppThemedTouchableOpacity Tests renders correctly with default props 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#CD03EF",
      "borderRadius": 10.714285714285714,
      "marginBottom": 21.428571428571427,
      "opacity": 1,
      "padding": 21.428571428571427,
      "width": "80%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#ECEDEE",
        },
        {
          "fontSize": 34.285714285714285,
          "lineHeight": 51.42857142857143,
        },
        undefined,
        undefined,
        undefined,
        undefined,
        [
          {
            "color": "#fff",
          },
          undefined,
        ],
      ]
    }
  >
    Click Me
  </Text>
</View>
`;
